src/Symfony/Component/Workflow/Workflow.php

```diff
@@ -0,0 +1,274 @@
+            return false;
+        }
+
+        return true;
+    }
+
+    private function guardTransition($subject, Marking $marking, Transition $transition)
+    {
+        if (null === $this->dispatcher) {
+            return;
+        }
+
+        $event = new GuardEvent($subject, $marking, $transition);
+
+        $this->dispatcher->dispatch('workflow.guard', $event);
+        $this->dispatcher->dispatch(sprintf('workflow.%s.guard', $this->name), $event);
+        $this->dispatcher->dispatch(sprintf('workflow.%s.guard.%s', $this->name, $transition->getName()), $event);
+
+        return $event->isBlocked();
+    }
+
+    private function leave($subject, Transition $transition, Marking $marking)
+    {
+        if (null !== $this->dispatcher) {
+            $event = new Event($subject, $marking, $transition);
+
+            $this->dispatcher->dispatch('workflow.leave', $event);
+            $this->dispatcher->dispatch(sprintf('workflow.%s.leave', $this->name), $event);
+        }
+
+        foreach ($transition->getFroms() as $place) {
+            $marking->unmark($place);
+
+            if (null !== $this->dispatcher) {
+                $this->dispatcher->dispatch(sprintf('workflow.%s.leave.%s', $this->name, $place), $event);
+            }
+        }
+    }
+
+    private function transition($subject, Transition $transition, Marking $marking)
+    {
+        if (null === $this->dispatcher) {
+            return;
+        }
+
+        $event = new Event($subject, $marking, $transition);
+
+        $this->dispatcher->dispatch('workflow.transition', $event);
+        $this->dispatcher->dispatch(sprintf('workflow.%s.transition', $this->name), $event);
+        $this->dispatcher->dispatch(sprintf('workflow.%s.transition.%s', $this->name, $transition->getName()), $event);
+    }
+
+    private function enter($subject, Transition $transition, Marking $marking)
+    {
+        if (null !== $this->dispatcher) {
+            $event = new Event($subject, $marking, $transition);
+
+            $this->dispatcher->dispatch('workflow.enter', $event);
+            $this->dispatcher->dispatch(sprintf('workflow.%s.enter', $this->name), $event);
+        }
+
+        foreach ($transition->getTos() as $place) {
+            $marking->mark($place);
+
+            if (null !== $this->dispatcher) {
+                $this->dispatcher->dispatch(sprintf('workflow.%s.enter.%s', $this->name, $place), $event);
+            }
+        }
+    }
+
+    private function announce($subject, Transition $initialTransition, Marking $marking)
+    {
+        if (null === $this->dispatcher) {
+            return;
+        }
+
+        $event = new Event($subject, $marking, $initialTransition);
+
+        foreach ($this->definition->getTransitions() as $transition) {
+            if ($this->doCan($subject, $marking, $transition)) {
+                $this->dispatcher->dispatch(sprintf('workflow.%s.announce.%s', $this->name, $transition->getName()), $event);
```

Comment:
Hello @lutangar ;

I'm not sure to understand what you want. Anyway, If you want a new feature, or to change an existing one, please open a new issue, it will be easier to discuss about it there. And feel free to ping me ;)